File: app/src/main/java/com/example/smarthome/relays/ui/AddRelayFragment.java
package com.example.smarthome.relays.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.example.smarthome.R;
import com.example.smarthome.relays.models.Relay;
import com.example.smarthome.relays.viewModels.RelayViewModel;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.switchmaterial.SwitchMaterial;

public class AddRelayFragment extends Fragment{

    private RelayViewModel viewModel;
    private ProgressBar progressBar;
    private FloatingActionButton fabAdd;
    private EditText etName;
    private EditText etIp;
    private SwitchMaterial switchOnOf;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        viewModel = new ViewModelProvider(requireActivity()).get(RelayViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.add_relay_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {

        etName = view.findViewById(R.id.relay_add_et_name);
        etIp = view.findViewById(R.id.relay_add_et_ip);
        switchOnOf = view.findViewById(R.id.relay_add_switch_onof);
        progressBar = view.findViewById(R.id.relay_add_pb);
        fabAdd = view.findViewById(R.id.relay_add_fab);

        initViewModelObservables();
        initOnClickListeners();
    }

    private void initViewModelObservables(){
        viewModel.getStatus().observe(getViewLifecycleOwner(), status -> {

            String message = "";

            switch(status.getStatus()){
                case LOADING:
                    progressBar.setVisibility(View.VISIBLE);
                    fabAdd.setVisibility(View.GONE);
                    break;
                default:
                    progressBar.setVisibility(View.GONE);
                    fabAdd.setVisibility(View.VISIBLE);
                    if(status.getMessage() != null) message = status.getMessage();

                    if(!message.isEmpty())
                        Toast.makeText(
                                requireActivity(), status.getMessage(), Toast.LENGTH_SHORT).show();
                    break;
            }
        });
    }

    private void initOnClickListeners(){
        fabAdd.setOnClickListener(v -> {
            String name = etName.getText().toString();
            String ip = etIp.getText().toString();
            boolean on = switchOnOf.isChecked();

            Relay relay = new Relay(name, ip, on);

            viewModel.add(relay);
        });
    }
}

File: app/src/main/java/com/example/smarthome/ui/relay/AddRelayFragment.kt
package com.example.smarthome.ui.relay

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import com.example.smarthome.R
import com.example.smarthome.data.model.Relay
import com.example.smarthome.databinding.AddRelayFragmentBinding
import com.example.smarthome.utilities.LiveDataObservers
import com.example.smarthome.utilities.OnClickListeners
import com.example.smarthome.utilities.Resource
import com.example.smarthome.utilities.ViewHelper
import com.example.smarthome.viewmodel.RelayViewModel
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class AddRelayFragment :
    Fragment(R.layout.add_relay_fragment),
    LiveDataObservers,
    OnClickListeners,
    ViewHelper
{
    private var _binding: AddRelayFragmentBinding? = null
    private val binding get() = _binding!!
    private val relayViewModel: RelayViewModel by viewModels()

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = AddRelayFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initLiveDataObservers()
        initOnClickListeners()
    }

    override fun initLiveDataObservers() {
        relayViewModel.status.observe(viewLifecycleOwner) {
            when (it.status){
                Resource.Status.LOADING -> setLoadingLayout()
                Resource.Status.ADDED -> requireActivity().finish()
                else -> setNormalLayout()
            }

            toast(requireActivity(), it.message)
        }
    }

    override fun initOnClickListeners() {
        binding.addButton.setOnClickListener {
            val name: String = binding.name.text.toString()
            val mac: String = binding.mac.text.toString()
            val isOn: Boolean = binding.switchButton.isChecked

            relayViewModel.add(Relay(name, mac, isOn))
        }
    }

    private fun setLoadingLayout() {
        showViews(binding.progressBar)
        hideViews(binding.addButton)
    }

    private fun setNormalLayout() {
        hideViews(binding.progressBar)
        showViews(binding.addButton)
    }
}
